client/src/connection/ssh/index.ts

```diff
@@ -25,175 +18,199 @@ export interface Config {
   privateKeyPath: string;
 }
 
-conn
-  .on("ready", () => {
-    conn.shell((err, s) => {
-      if (err) {
-        reject?.(err);
-        return;
-      }
-      stream = s;
-      if (!stream) {
-        reject?.(err);
+export function getSession(c: Config): Session {
+  if (!sessionInstance) {
+    sessionInstance = new SSHSession(c);
+  }
+  return sessionInstance;
+}
+
+export class SSHSession implements Session {
+  private conn: Client;
+  private stream: ClientChannel | undefined;
+  private config: Config;
+  private resolve: ((value?) => void) | undefined;
+  private reject: ((reason?) => void) | undefined;
+  private onLog: ((logs: LogLine[]) => void) | undefined;
+  private logs: string[] = [];
+  private html5FileName = "";
+  private timer: NodeJS.Timeout;
+
+  constructor(c: Config) {
+    this.config = c;
+    this.conn = new Client();
+  }
+
+  setup = (): Promise<void> => {
+    return new Promise((pResolve, pReject) => {
+      this.resolve = pResolve;
+      this.reject = pReject;
+
+      if (this.stream) {
+        this.resolve();
         return;
       }
 
-      stream
-        .on("close", () => {
-          onLog = undefined;
-          stream = undefined;
-          resolve = undefined;
-          reject = undefined;
-          logs = [];
-          conn.end();
-        })
-        .on("data", (data: Buffer) => {
-          const output = data.toString().trimEnd();
-          const outputLines = output.split(/\n|\r\n/);
-          if (onLog) {
-            outputLines.forEach((line) => {
-              if (!line) {
-                return;
-              }
-              if (line.endsWith(endCode)) {
-                // run completed
-                getResult();
-              }
-              if (!(line.endsWith("?") || line.endsWith(">"))) {
-                html5FileName =
-                  line.match(/NOTE: .+ HTML5.* Body .+: (.+)\.htm/)?.[1] ??
-                  html5FileName;
-                onLog?.([{ type: "normal", line }]);
+      const cfg: ConnectConfig = {
+        host: this.config.host,
+        port: this.config.port,
+        username: this.config.username,
+        privateKey: readFileSync(this.config.privateKeyPath),
+        readyTimeout: sasLaunchTimeout,
+      };
+
+      this.conn
+        .on("ready", () => {
+          this.conn.shell((err, s) => {
+            if (err) {
+              this.reject?.(err);
+              return;
+            }
+            this.stream = s;
+            if (!this.stream) {
+              this.reject?.(err);
+              return;
+            }
+
+            this.stream.on("close", () => {
+              this.onLog = undefined;
+              this.stream = undefined;
+              this.resolve = undefined;
+              this.reject = undefined;
+              this.logs = [];
+              this.html5FileName = "";
+              this.timer = undefined;
+              this.conn.end();
```

Comment:
What's happening here? Is this "resetting" or "clearing" our session? If so, I might suggest creating a private method to handle that. Then your code could be something like...
```
this.stream.on('close', this.reset)
```

Reply:
So it turns out that getSession() is called for pretty much all 3 phases (setup, run close) of code execution when dealing with sessions. This is attempting to re-initialize the instance fields when the user closes the connection since the instance itself will get reused (due to the getSession item I mention above). Will move some of these heavier blocks of logic off into private methods to make it easier to reason about what's going on.